src/dimension.c

```diff
@@ -957,8 +957,84 @@ ts_point_create(int16 num_dimensions)
 	return p;
 }
 
+int
+get_partition_attrno(Dimension *d, TupleTableSlot *slot, List *colnames_list)
+{
+	AttrNumber partition_col_idx = 0;
+	bool partition_col_found = false;
+	if (colnames_list)
+	{
+		ListCell *col_list;
+		foreach (col_list, colnames_list)
+		{
+			const char *col_name = lfirst(col_list);
+			if (namestrcmp(&d->fd.column_name, col_name) == 0)
+			{
+				partition_col_found = true;
+				break;
+			}
+			partition_col_idx++;
+		}
+	}
+	if (partition_col_found)
+	{
+		TupleDesc tupdesc = slot->tts_tupleDescriptor;
+		if (tupdesc->attrs[partition_col_idx].atttypid == d->fd.column_type)
+			return partition_col_idx + 1;
+	}
+	return 0;
+}
+
+/*
+ * Helper function to find the right partition value from a tuple,
+ * in case attributes in tuple are not ordered correctly.
+ *
+ * Assume columns in table are in below order:
+ *  [time, temperature, location, value]
+ * "time" is the partition column based on this columns value
+ * ChunkDispatch node decides if to create or reuse an existing
+ * chunk.
```

Comment:
I think we need to have a test with the specific configuration you mention here.

Reply:
I do have a testcase for this in merge.sql
Please notice the column order in table vs the column order in the tuple received from child node of ChunkDispatch.

```
postgres=# \d target
                              Table "public.target"
   Column    |           Type           | Collation | Nullable |     Default
-------------+--------------------------+-----------+----------+------------------
 time        | timestamp with time zone |           | not null |
 location    | smallint                 |           | not null |
 temperature | double precision         |           |          |
 val         | text                     |           |          | 'string -'::text
Indexes:
    "target_time_idx" btree ("time" DESC)
Triggers:
    ts_insert_blocker BEFORE INSERT ON target FOR EACH ROW EXECUTE FUNCTION _timescaledb_internal.insert_blocker()
Number of child tables: 2 (Use \d+ to list them.)

postgres=#
postgres=# explain (costs off, verbose) MERGE INTO target t                                                                                                                                                                                USING source s                                                                                                                                                                                                                             ON t.time = s.time AND t.location = s.location                                                                                                                                                                                             WHEN MATCHED THEN                                                                                                                                                                                                                          UPDATE SET temperature = (t.temperature + s.temperature)/2, val = val || ' UPDATED BY MERGE'                                                                                                                                               WHEN NOT MATCHED THEN                                                                                                                                                                                                                      INSERT (time, location, temperature, val) VALUES (s.time, s.location, s.temperature, 'string - INSERTED BY MERGE');
                                           QUERY PLAN
------------------------------------------------------------------------------------------------
 Custom Scan (HypertableModify)
   ->  Merge on public.target t
         Merge on public.target t_1
         Merge on _timescaledb_internal._hyper_38_13_chunk t_2
         Merge on _timescaledb_internal._hyper_38_14_chunk t_3
         ->  Custom Scan (ChunkDispatch)
               ->  Merge Left Join
                     Output: s.temperature, s."time", s.location, t.tableoid, t.ctid
                     Merge Cond: ((s."time" = t."time") AND (s.location = t.location))
                     ->  Sort
                           Output: s.temperature, s."time", s.location
                           Sort Key: s."time", s.location
                           ->  Append
                                 ->  Seq Scan on public.source s_1
                                       Output: s_1.temperature, s_1."time", s_1.location
                                 ->  Seq Scan on _timescaledb_internal._hyper_39_15_chunk s_2
                                       Output: s_2.temperature, s_2."time", s_2.location
                                 ->  Seq Scan on _timescaledb_internal._hyper_39_16_chunk s_3
                                       Output: s_3.temperature, s_3."time", s_3.location
                     ->  Sort
                           Output: t."time", t.location, t.tableoid, t.ctid
                           Sort Key: t."time", t.location
                           ->  Append
                                 ->  Seq Scan on public.target t_1
                                       Output: t_1."time", t_1.location, t_1.tableoid, t_1.ctid
                                 ->  Seq Scan on _timescaledb_internal._hyper_38_13_chunk t_2
                                       Output: t_2."time", t_2.location, t_2.tableoid, t_2.ctid
                                 ->  Seq Scan on _timescaledb_internal._hyper_38_14_chunk t_3
                                       Output: t_3."time", t_3.location, t_3.tableoid, t_3.ctid
(29 rows)
```